File: js/merge-sort.js
let arr = [2,24,24,4,56,7,24,542,547,698,80,17,1,29,48,29,37,64,267,29,584,27,25,45,475,6,8862,52,65,86,37,2,4,6,8,0,4,3,2,1,45,7];

const mergeSort = (arr)=>{

  const arrSize = arr.length;
  // base case
  if(arrSize < 2){
    return
  }
  const middle = Math.trunc(arrSize / 2);
  let ls = arr.slice(0, middle);
  let rs = arr.slice(middle, arrSize);

  // recursion
  mergeSort(ls);
  mergeSort(rs);

  // Merge
  let i=0 ,j=0 ,k = 0;

  while(i < ls.length && j < rs.length) {
    if (ls[i] < rs[j]) {
      arr[k] = ls[i];
      i++;
    } else {
      arr[k] = rs[j];
      j++;
    }
    k++;
  }
  while(i < ls.length) {
    arr[k] = ls[i];
    i++;
    k++;
  }
  while(j < rs.length) {
    arr[k] = rs[j];
    j++;
    k++;
  }
}
console.log(arr);
mergeSort(arr);
console.log(arr);

File: python/merge-sort.py
arr = [2,24,25,47,58,326,125,412,25,47,53,950,569,46,24,14,16,18,209,27,47,59,40,27,38,49,450,37,78,679,8,236]

def merge(ls, rs):
    sorted_arr = []
    while len(ls) > 0 and len(rs) > 0:
        if ls[0] < rs[0]:
            sorted_arr.append(ls[0])
            ls.pop(0)
        else :
            sorted_arr.append(rs[0])
            rs.pop(0)
    while len(ls) > 0:
        sorted_arr.append(ls[0])
        ls.pop(0)

    while len(rs) > 0:
        sorted_arr.append(rs[0])
        rs.pop(0)

    return sorted_arr
def merge_sort(array):
    if len(array) == 1:
        return array

    middle = len(array) // 2
    ls = array[:middle]
    rs = array[middle:]

    sorted_ls = merge_sort(ls)
    sorted_rs = merge_sort(rs)

    return merge(sorted_ls, sorted_rs)

def main():
    is_sorted = merge_sort(arr)
    print(is_sorted)

if __name__ == "__main__":
    main()



    # [2,4,5,7,8,9,] [1,7,8,9,10,29]

